libraries/AC_AutoTune/AC_AutoTune_Multi.cpp

```diff
@@ -0,0 +1,626 @@
+#define AUTOTUNE_SP_STEP                  0.05f     // minimum increment when increasing/decreasing Stab P term
+#define AUTOTUNE_PI_RATIO_FOR_TESTING      0.1f     // I is set 10x smaller than P during testing
+#define AUTOTUNE_RD_MAX                  0.200f     // maximum Rate D value
+#define AUTOTUNE_RLPF_MIN                  1.0f     // minimum Rate Yaw filter value
+#define AUTOTUNE_RLPF_MAX                  5.0f     // maximum Rate Yaw filter value
+#define AUTOTUNE_RP_MIN                   0.01f     // minimum Rate P value
+#define AUTOTUNE_RP_MAX                    2.0f     // maximum Rate P value
+#define AUTOTUNE_SP_MAX                   20.0f     // maximum Stab P value
+#define AUTOTUNE_SP_MIN                    0.5f     // maximum Stab P value
+#define AUTOTUNE_D_UP_DOWN_MARGIN          0.2f     // The margin below the target that we tune D in
+
+#include "AC_AutoTune_Multi.h"
+
+void AC_AutoTune_Multi::do_gcs_announcements()
+{
+    const uint32_t now = AP_HAL::millis();
+    if (now - announce_time < AUTOTUNE_ANNOUNCE_INTERVAL_MS) {
+        return;
+    }
+    float tune_rp = 0.0f;
+    float tune_rd = 0.0f;
+    float tune_sp = 0.0f;
+    float tune_accel = 0.0f;
+    char axis_char = '?';
+    switch (axis) {
+    case ROLL:
+        tune_rp = tune_roll_rp;
+        tune_rd = tune_roll_rd;
+        tune_sp = tune_roll_sp;
+        tune_accel = tune_roll_accel;
+        axis_char = 'R';
+        break;
+    case PITCH:
+        tune_rp = tune_pitch_rp;
+        tune_rd = tune_pitch_rd;
+        tune_sp = tune_pitch_sp;
+        tune_accel = tune_pitch_accel;
+        axis_char = 'P';
+        break;
+    case YAW:
+        tune_rp = tune_yaw_rp;
+        tune_rd = tune_yaw_rLPF;
+        tune_sp = tune_yaw_sp;
+        tune_accel = tune_yaw_accel;
+        axis_char = 'Y';
+        break;
+    }
+
+    gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: (%c) %s", axis_char, type_string());
+    send_step_string();
+    if (!is_zero(lean_angle)) {
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: lean=%f target=%f", (double)lean_angle, (double)target_angle);
+    }
+    if (!is_zero(rotation_rate)) {
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: rotation=%f target=%f", (double)(rotation_rate*0.01f), (double)(target_rate*0.01f));
+    }
+    switch (tune_type) {
+    case RD_UP:
+    case RD_DOWN:
+    case RP_UP:
+    case RP_DOWN:
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: p=%f d=%f", (double)tune_rp, (double)tune_rd);
+        break;
+    case RFF_UP:
+    case RFF_DOWN:
+        break;
+    case SP_DOWN:
+    case SP_UP:
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: p=%f accel=%f", (double)tune_sp, (double)tune_accel);
+        break;
+    case MAX_GAINS:
+    case TUNE_COMPLETE:
+        break;
+    }
+    gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: success %u/%u", counter, AUTOTUNE_SUCCESS_COUNT);
+
+    announce_time = now;
+}
+
+void AC_AutoTune_Multi::test_init()
+{
+
```

Comment:
no big deal but normally we wouldn't have these extra blank lines above and below, "twitch_test_init()".  same for "test_run()"

Reply:
done.  Thanks!